### PR TITLE
refactor[next] Prepare new Field for itir.embedded

### PR DESCRIPTION
**Pull request by havogt**
- improve TypeAliases
- add `domain` and `unit_range` constructors
- extract domain slicing utils to `next.embedded.common`
- introduce `MutableField`
- add some missing operators to `Field`

### CODE REVIEW

File: src/gt4py/next/embedded/nd_array_field.py [hunk 1]
```diff
@@ -194,35 +221,58 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
+        raise NotImplementedError("`__xor__` not implemented for non-`bool` fields.")
+
+    __rxor__ = __xor__
+
+    def __invert__(self) -> _BaseNdArrayField:
+        if self.dtype == core_defs.BoolDType():
+            return _make_unary_array_field_intrinsic_func("invert", "invert")(self)
+        raise NotImplementedError("`__invert__` not implemented for non-`bool` fields.")
+
+    def _slice(self, index: common.FieldSlice) -> tuple[common.Domain, common.BufferSlice]:
```
Comment (samkellerhals) [inline src/gt4py/next/embedded/nd_array_field.py:256]: It may be useful to have `_slice` as a utility method so that it can be used by other fields such as `ConstantField` as the slicing methodology is shared.

File: src/gt4py/next/common.py [hunk 2]
```diff
@@ -90,6 +90,17 @@ def __post_init__(self):
             object.__setattr__(self, "start", 0)
             object.__setattr__(self, "stop", 0)
 
+    @classmethod
+    def from_unit_range_like(cls, r: UnitRangeLike) -> UnitRange:
```
Comment (havogt) [inline src/gt4py/next/common.py:94]: make free function `unit_range(r: UnitRangeLike) -> UnitRange`

File: src/gt4py/next/common.py [hunk 3]
```diff
@@ -136,34 +147,129 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+            isinstance(e, tuple) and isinstance(e[0], Dimension) and is_unit_range_like(e[1])
+            for e in v
+        )
+    ) or (
+        isinstance(v, Mapping)
+        and all(isinstance(d, Dimension) and is_unit_range_like(r) for d, r in v.items())
+    )
+
+
+def to_named_range(v: tuple[Dimension, UnitRangeLike]) -> NamedRange:
```
Comment (havogt) [inline src/gt4py/next/common.py:214]:
```suggestion
def named_range(v: tuple[Dimension, UnitRangeLike]) -> NamedRange:
```

File: src/gt4py/next/common.py [hunk 4]
```diff
@@ -136,34 +147,129 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+            object.__setattr__(self, "ranges", tuple(ranges))
+
         if len(set(self.dims)) != len(self.dims):
             raise NotImplementedError(f"Domain dimensions must be unique, not {self.dims}.")
 
-        if len(self.dims) != len(self.ranges):
-            raise ValueError(
-                f"Number of provided dimensions ({len(self.dims)}) does not match number of provided ranges ({len(self.ranges)})."
+    @classmethod
+    def from_domain_like(cls, domain_like: DomainLike) -> Domain:
```
Comment (havogt) [inline src/gt4py/next/common.py:256]: free function `domain`

File: tests/next_tests/unit_tests/test_common.py [hunk 5]
```diff
@@ -305,3 +344,5 @@ def test_dimension_promotion(
             promote_dims(*dim_list)
 
         assert exc_info.match(expected_error_msg)
+        assert exc_info.match(expected_error_msg)
+        assert exc_info.match(expected_error_msg)
```
Comment (havogt) [inline tests/next_tests/unit_tests/test_common.py:348]: remove accidently duplicated..

File: src/gt4py/next/common.py [hunk 6]
```diff
@@ -205,15 +292,33 @@ def __and__(self, other: "Domain") -> "Domain":
                 _broadcast_ranges(broadcast_dims, other.dims, other.ranges),
             )
         )
-        return Domain(broadcast_dims, intersected_ranges)
+        return Domain(dims=broadcast_dims, ranges=intersected_ranges)
+
+
+def domain(domain_like: DomainLike) -> Domain:
```
Comment (havogt) [inline src/gt4py/next/common.py:298]: doctest

File: src/gt4py/next/common.py [hunk 7]
```diff
@@ -467,17 +590,10 @@ def promote_dims(*dims_list: Sequence[Dimension]) -> list[Dimension]:
...
-    return isinstance(index, Sequence) and all(
-        is_named_range(idx) or is_named_index(idx) for idx in index
-    )
+    return topologically_sorted_list
+    return topologically_sorted_list
+    return topologically_sorted_list
+    return topologically_sorted_list
+    return topologically_sorted_list
+    return topologically_sorted_list
+    return topologically_sorted_list
```
Comment (havogt) [inline src/gt4py/next/common.py:599]: wtf

File: src/gt4py/next/embedded/common.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+def _find_index_of_dim(
+    dim: common.Dimension,
+    domain_slice: common.Domain | Sequence[common.NamedRange | common.NamedIndex | Any],
+) -> Optional[int]:
+    for i, (d, _) in enumerate(domain_slice):
+        if dim == d:
+            return i
+    return None
+    return None
```
Comment (havogt) [inline src/gt4py/next/embedded/common.py:145]: wtf

File: src/gt4py/next/embedded/nd_array_field.py [hunk 9]
```diff
@@ -421,34 +438,8 @@ def _compute_slice(rng: common.DomainRange, domain: common.Domain, pos: int) ->
...
-    indices: tuple[int | slice | EllipsisType, ...], target_size: int
-) -> tuple[int | slice, ...]:
-    expanded_indices: list[int | slice] = []
-    for idx in indices:
-        if idx is Ellipsis:
-            expanded_indices.extend([slice(None)] * (target_size - (len(indices) - 1)))
-        else:
-            expanded_indices.append(idx)
-    return tuple(expanded_indices)
+        raise ValueError(f"Can only use integer or UnitRange ranges, provided type: {type(rng)}")
```
Comment (havogt) [inline src/gt4py/next/embedded/nd_array_field.py:445]: wtf

File: src/gt4py/next/embedded/nd_array_field.py (hunk 1 above)
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:257]: done

File: src/gt4py/next/embedded/common.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+    start = (
+        input_range.start if slice_obj.start is None or slice_obj.start >= 0 else input_range.stop
+    ) + (slice_obj.start or 0)
+    stop = (
+        input_range.start if slice_obj.stop is None or slice_obj.stop >= 0 else input_range.stop
+    ) + (slice_obj.stop or len(input_range))
+
+    if start < input_range.start or stop > input_range.stop:
+        raise IndexError()
```
Comment (egparedes) [inline src/gt4py/next/embedded/common.py:132]: Missing exception message

File: src/gt4py/next/embedded/nd_array_field.py [hunk 11]
```diff
@@ -148,38 +162,52 @@ def from_array(
...
     def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
         raise NotImplementedError()
 
+    def restrict(self, index: common.FieldSlice) -> common.Field | core_defs.ScalarT:
+        new_domain, buffer_slice = self._slice(index)
+
+        new_buffer = self.ndarray[buffer_slice]
+        if len(new_domain) == 0:
+            assert core_defs.is_scalar_type(new_buffer)
+            return new_buffer  # type: ignore[return-value] # I don't think we can express that we return `ScalarT` here
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:197]: Should we change the behavior to always return fields as in the DataArray API standard in this PR? What about always requiring `Ellipsis` in relative slicing?

File: src/gt4py/next/embedded/nd_array_field.py [hunk 12]
```diff
@@ -194,78 +222,50 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
-            and not common.is_domain_slice(index)
-            or common.is_named_index(index)
-            or common.is_named_range(index)
-        ):
-            index = cast(common.FieldSlice, (index,))
+    __mod__ = __rmod__ = _make_binary_array_field_intrinsic_func("mod", "mod")
 
-        if common.is_domain_slice(index):
-            return self._getitem_absolute_slice(index)
+    def __and__(self, other: common.Field) -> _BaseNdArrayField:
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:227]: We should also add it to the `Field` protocol if it's not there.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 13]
```diff
@@ -194,78 +222,50 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
+            )
+        raise NotImplementedError("`__and__` not implemented for non-`bool` fields.")
 
-        assert isinstance(index, tuple)
-        if all(isinstance(idx, (slice, int)) or idx is Ellipsis for idx in index):
-            return self._getitem_relative_slice(index)
+    __rand__ = __and__
 
-        raise IndexError(f"Unsupported index type: {index}")
+    def __or__(self, other: common.Field) -> _BaseNdArrayField:
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:236]: We should also add it to the `Field` protocol if it's not there.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 14]
```diff
@@ -194,78 +222,50 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
-                if isinstance(index_or_range, common.UnitRange):
-                    new_ranges.append(index_or_range)
-                    new_dims.append(dim)
-            else:
-                # dimension not mentioned in slice
-                new_ranges.append(self.domain.ranges[i])
-                new_dims.append(dim)
-
-        new_domain = common.Domain(dims=tuple(new_dims), ranges=tuple(new_ranges))
+    def __xor__(self, other: common.Field) -> _BaseNdArrayField:
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:243]: We should also add it to the `Field` protocol if it's not there.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 15]
```diff
@@ -194,78 +222,50 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
-
-        new_domain = common.Domain(dims=tuple(new_dims), ranges=tuple(new_ranges))
+    __rxor__ = __xor__
 
-        if len(new_domain) == 0:
-            assert core_defs.is_scalar_type(new), new
-            return new  # type: ignore[return-value] # I don't think we can express that we return `ScalarT` here
-        else:
-            return self.__class__.from_array(new, domain=new_domain, value_type=self.value_type)
+    def __invert__(self) -> _BaseNdArrayField:
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:252]: We should also add it to the `Field` protocol if it's not there.

File: src/gt4py/next/embedded/nd_array_field.py [hunk 16]
```diff
@@ -194,78 +222,50 @@ def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
...
+        new_domain = embedded_common.sub_domain(self.domain, index)
+
+        index = embedded_common._tuplize_field_slice(index)
+
+        slice_ = (
+            _get_slices_from_domain_slice(self.domain, index)
+            if common.is_domain_slice(index)
+            else index
+        )
+        assert common.is_buffer_slice(slice_), slice_
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:267]: Leftover from debugging sessions? Otherwise let's add a small error message

File: src/gt4py/next/embedded/nd_array_field.py [hunk 17]
```diff
@@ -325,38 +348,30 @@ class CuPyArrayField(_BaseNdArrayField):
...
         array_ns: ClassVar[ModuleType] = jnp
 
-    common.field.register(jnp.ndarray, JaxArrayField.from_array)
-
+        def __setitem__(
+            self,
+            index: common.FieldSlice,
+            value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+        ) -> None:
+            # use `self.ndarray.at(index).set(value)`
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:356]:
```suggestion
            # TODO: use something like `self.ndarray. = self.ndarray.at(index).set(value)`
```

File: src/gt4py/next/embedded/nd_array_field.py [hunk 18]
```diff
@@ -16,16 +16,16 @@
...
-from typing import Any, ClassVar, Optional, ParamSpec, TypeAlias, TypeVar, cast, overload
+from types import ModuleType
+from typing import Any, ClassVar, Optional, ParamSpec, TypeAlias, TypeVar, overload
 
 import numpy as np
 from numpy import typing as npt
 
 from gt4py._core import definitions as core_defs
 from gt4py.next import common
+from gt4py.next.embedded import common as embedded_common
```
Comment (egparedes) [inline src/gt4py/next/embedded/nd_array_field.py:28]: Really ugly alias. Would you consider something different like  `emb_common`, `e_common`, `embedded_co`, ...?

File: tests/next_tests/unit_tests/embedded_tests/test_common.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+from gt4py.next.embedded import exceptions as embedded_exceptions
+from gt4py.next.embedded.common import _slice_range, sub_domain
+
+
+def test_slice_range():
+    input_range = UnitRange(2, 10)
+    slice_obj = slice(2, -2)
+    expected = UnitRange(4, 8)
+
+    result = _slice_range(input_range, slice_obj)
```
Comment (egparedes) [inline tests/next_tests/unit_tests/embedded_tests/test_common.py:30]: I would add a test for `slice(None)`

**Review by egparedes — CHANGES_REQUESTED** [review]
The review of the PR has been more complicated than expected because the PR is not small and the changes are not simple. I have many comments, many of them are simple python coding style remarks but there are also some questions and comments about the logic, which I hope I didn't get wrong.

Still need to review the tests in more depth in the next iteration.

File: src/gt4py/next/common.py [hunk 20]
```diff
@@ -64,17 +64,23 @@ class DimensionKind(StrEnum):
...
+        return self.value
+
 
 @dataclasses.dataclass(frozen=True)
 class Dimension:
     value: str
     kind: DimensionKind = dataclasses.field(default=DimensionKind.HORIZONTAL)
 
+    def __repr__(self):
+        return f'Dimension(value="{self.value}", kind={repr(self.kind)})'
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: I guess I copied from somewhere, anyway deleted now as the default repr works.

File: src/gt4py/next/__init__.py [hunk 21]
```diff
@@ -25,6 +25,7 @@
 
 from . import common, ffront, iterator, program_processors, type_inference
 from .common import Dimension, DimensionKind, Field, GridType
+from .embedded import nd_array_field
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/__init__.py:28]: because otherwise nd_array_field is not registered by default in `common.field` dispatcher

File: src/gt4py/next/embedded/common.py [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+from gt4py.next.embedded import exceptions as embedded_exceptions
+
+
+def sub_domain(domain: common.Domain, index: common.FieldSlice) -> common.Domain:
+    index = _tuplize_field_slice(index)
+
+    if common.is_domain_slice(index):
+        return _absolute_sub_domain(domain, index)
+
+    assert isinstance(index, tuple)
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/common.py:29]: if is_any_index(...):
    index = (index, )

File: src/gt4py/next/embedded/common.py [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+                )
+        else:
+            # not in new domain
+            assert common.is_int_index(idx)
+            new_index = (rng.start if idx >= 0 else rng.stop) + idx
+            if new_index < rng.start or new_index >= rng.stop:
+                raise embedded_exceptions.IndexOutOfBounds(
+                    domain=domain, indices=index, index=idx, dim=dim
+                )
+
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/common.py:61]: reduce dimension

File: src/gt4py/next/common.py [hunk 24]
```diff
@@ -135,36 +141,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
         else:
             raise NotImplementedError("Can only find the intersection between UnitRange instances.")
 
+    def __str__(self) -> str:
+        return f"({self.start}:{self.stop})"
 
-DomainRange: TypeAlias = UnitRange | int
+
+def unit_range(r: UnitRangeLike) -> UnitRange:
+    assert is_unit_range_like(r)
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:143]: I removed the `UnitRangeLike` as a type completely and just kept `RangeLike`. So unit_range takes `RangelLike` and gives error in case it cannot be interpreted as UnitRange.

File: src/gt4py/next/common.py [hunk 25]
```diff
@@ -135,36 +141,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+    if isinstance(r, UnitRange):
+        return r
+    if isinstance(r, range) and r.step == 1:
+        return UnitRange(r.start, r.stop)
+    if isinstance(r, tuple) and isinstance(r[0], int) and isinstance(r[1], int):
+        return UnitRange(r[0], r[1])
+    raise ValueError(f"`{r}` is not `UnitRangeLike`.")
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:150]: see comment above.

File: src/gt4py/next/common.py [hunk 26]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+FieldSlice: TypeAlias = DomainSlice | BufferSlice | AnyIndex
+UnitRangeLike: TypeAlias = UnitRange | range | tuple[int, int]
+DomainLike: TypeAlias = (
+    Sequence[tuple[Dimension, UnitRangeLike]] | Mapping[Dimension, UnitRangeLike]
 )
 
 
-@dataclasses.dataclass(frozen=True)
+def is_int_index(p: Any) -> TypeGuard[IntIndex]:
+    return isinstance(p, (int, np.integer))
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:168]: It is because np.integer is not a simple type, I get
```
src/gt4py/next/common.py:186:13: error: Parameterized generics cannot be used with class or instance checks  [misc]
src/gt4py/next/common.py:186:30: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
```
Probably we should not use np.integer, but `core_defs.IntegralScalar`?

[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:168]: but it has the same problem

File: src/gt4py/next/common.py [hunk 27]
```diff
@@ -135,36 +141,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+        return r
+    if isinstance(r, range) and r.step == 1:
+        return UnitRange(r.start, r.stop)
+    if isinstance(r, tuple) and isinstance(r[0], int) and isinstance(r[1], int):
+        return UnitRange(r[0], r[1])
+    raise ValueError(f"`{r}` is not `UnitRangeLike`.")
+
+
+IntIndex: TypeAlias = int | np.integer
+DomainRange: TypeAlias = UnitRange | IntIndex
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:154]: I don't think there is a good name, I'll remove it and inline it in the single place where it's used

File: src/gt4py/next/common.py [hunk 28]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+    return isinstance(p, (int, np.integer))
+
+
+def is_named_range(v: Any) -> TypeGuard[NamedRange]:
+    return (
+        isinstance(v, tuple)
+        and len(v) == 2
+        and isinstance(v[0], Dimension)
+        and isinstance(v[1], UnitRange)
+    )
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:177]: constrained to AnyIndex

File: src/gt4py/next/common.py [hunk 29]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+
+
+def is_named_index(v: Any) -> TypeGuard[NamedRange]:
+    return (
+        isinstance(v, tuple) and len(v) == 2 and isinstance(v[0], Dimension) and is_int_index(v[1])
+    )
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:183]: see above

File: src/gt4py/next/common.py [hunk 30]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+
+
+def is_domain_slice(v: Any) -> TypeGuard[DomainSlice]:
+    return isinstance(v, Sequence) and all(is_named_range(e) or is_named_index(e) for e in v)
+
+
+def is_buffer_slice(v: Any) -> TypeGuard[BufferSlice]:
+    return isinstance(v, tuple) and all(
+        isinstance(e, slice) or is_int_index(e) or e is Ellipsis for e in v
+    )
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:193]: constrained to `AnyIndexSequence`

File: src/gt4py/next/common.py [hunk 31]
```diff
@@ -205,15 +297,50 @@ def __and__(self, other: "Domain") -> "Domain":
...
+    ):
+        return Domain(*tuple(named_range(d) for d in domain_like))
+    if isinstance(domain_like, Mapping) and all(
+        isinstance(d, Dimension) and is_unit_range_like(r) for d, r in domain_like.items()
+    ):
+        return Domain(
+            dims=tuple(domain_like.keys()),
+            ranges=tuple(unit_range(r) for r in domain_like.values()),
+        )
+    raise ValueError(f"`{domain_like}` is not `DomainLike`.")
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:336]: I changed the pattern in a different way, the responsibility for user errors is moved to `Domain.__init__`

File: src/gt4py/next/common.py [hunk 32]
```diff
@@ -325,29 +461,38 @@ def __pow__(self, other: Field | core_defs.ScalarT) -> Field:
...
 
-    @final
-    def __setattr__(self, key, value) -> None:
-        raise TypeError("Immutable type")
 
-    @final
-    def __setitem__(self, key, value) -> None:
-        raise TypeError("Immutable type")
+def is_mutable_field(
+    v: Any,
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:480]: done

File: src/gt4py/next/common.py [hunk 33]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
-FieldSlice: TypeAlias = (
-    DomainSlice
-    | tuple[slice | int | EllipsisType, ...]
-    | slice
-    | int
-    | EllipsisType
-    | NamedRange
-    | NamedIndex
+BufferSlice: TypeAlias = tuple[slice | IntIndex | EllipsisType, ...]
+FieldSlice: TypeAlias = DomainSlice | BufferSlice | AnyIndex
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:160]: changed in the meantime as discussed

File: src/gt4py/next/embedded/common.py [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+
+def sub_domain(domain: common.Domain, index: common.FieldSlice) -> common.Domain:
+    index = _tuplize_field_slice(index)
+
+    if common.is_domain_slice(index):
+        return _absolute_sub_domain(domain, index)
+
+    assert isinstance(index, tuple)
+    if all(isinstance(idx, slice) or common.is_int_index(idx) or idx is Ellipsis for idx in index):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/common.py:30]: fixed

File: src/gt4py/next/embedded/common.py (hunk 10 above)
Comment (havogt) [reply src/gt4py/next/embedded/common.py:132]: The only purpose is to catch the exception at the only call site. I don't like the pattern, but nobody should ever see the exception.

File: src/gt4py/next/embedded/nd_array_field.py (hunk 11 above)
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:197]: No, not in this PR, or we need to have a chat first to figure out how to extract elements (because this PR is preparing itir embedded, which requires it)

File: src/gt4py/next/embedded/nd_array_field.py (hunk 18 above)
Comment (havogt) [reply src/gt4py/next/embedded/nd_array_field.py:28]: I don't like any including the verbose, if you want a change, you have to pick.

File: src/gt4py/next/common.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:168]: resolved as discussed

File: src/gt4py/next/common.py [hunk 35]
```diff
@@ -135,36 +135,128 @@ def __and__(self, other: Set[Any]) -> UnitRange:
...
+
+
+def is_named_index(v: Any) -> TypeGuard[NamedRange]:
+    return (
+        isinstance(v, tuple) and len(v) == 2 and isinstance(v[0], Dimension) and is_int_index(v[1])
+    )
+
+
+def is_domain_slice(v: Any) -> TypeGuard[DomainSlice]:
+    return isinstance(v, Sequence) and all(is_named_range(e) or is_named_index(e) for e in v)
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:187]: constrained to `AnyIndexSequence`

File: src/gt4py/next/embedded/common.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/common.py:29]: done

File: src/gt4py/next/embedded/common.py [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+
+def _relative_sub_domain(domain: common.Domain, index: common.BufferSlice) -> common.Domain:
+    named_ranges: list[common.NamedRange] = []
+
+    expanded = _expand_ellipsis(index, len(domain))
+    if len(domain) < len(expanded):
+        raise IndexError(f"Trying to index a `Field` with {len(domain)} dimensions with {index}.")
+    for (dim, rng), idx in itertools.zip_longest(  # type: ignore[misc] # "slice" object is not iterable, not sure which slice...
+        domain, expanded, fillvalue=slice(None)
+    ):
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/embedded/common.py]: changed as suggested

**Review by egparedes — COMMENTED** [review]
It looks good to me, just minor suggestions. Let me know if you're fully done and I'll approve it

File: src/gt4py/next/__init__.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/__init__.py:28]: Ok, I see now. But I think that for coherency we should add `embedded` (and other missing submodules) to the list of imported submodules in line 26 and make sure that they are all reexported, or just remove the imports in line 26 and add a comment here saying that this is just imported for triggering the registering hook, but it shouldn't be used directly from here:
```suggestion
from .embedded import nd_array_field as _nd_array_field  # Just for registering field implementations
```

[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/__init__.py:28]: > to the list of imported submodules in line 26 and make sure that they are all reexported

What's the idea behind that?

File: src/gt4py/next/common.py [hunk 37]
```diff
@@ -231,7 +236,8 @@ def __init__(
         *args: NamedRange,
         dims: Optional[tuple[Dimension, ...]] = None,
         ranges: Optional[tuple[UnitRange, ...]] = None,
-    ):
+    ) -> None:
+        # TODO throw user error in case pre-conditions are not met
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:240]: yes, thanks.

File: src/gt4py/next/common.py [hunk 38]
```diff
@@ -304,6 +332,9 @@ def __str__(self) -> str:
         return f"Domain({', '.join(f'{e[0]}={e[1]}' for e in self)})"
 
 
+DomainLike: TypeAlias = Sequence[tuple[Dimension, RangeLike]] | Mapping[Dimension, RangeLike]
+
+
 def domain(domain_like: DomainLike) -> Domain:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:338]: `Domain` is a `Sequence[NamedRange]` therefore a subset of  `Sequence[tuple[Dimension, RangeLike]]`, should I add a comment?

File: src/gt4py/next/common.py [hunk 39]
```diff
@@ -613,3 +654,26 @@ def promote_dims(*dims_list: Sequence[Dimension]) -> list[Dimension]:
...
+
+
+class FieldBuiltinFuncRegistry:
+    _builtin_func_map: collections.ChainMap[
+        fbuiltins.BuiltInFunction, Callable
+    ] = collections.ChainMap()
+
+    def __init_subclass__(cls, **kwargs):
+        # might break in multiple inheritance (if multiple ancestors have `_builtin_func_map`)
+        cls._builtin_func_map = cls._builtin_func_map.new_child()
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:666]: What's the `[0]` in `c.__dict__["_builtin_func_map"][0]`?

File: src/gt4py/next/common.py [hunk 40]
```diff
@@ -140,36 +142,42 @@ def __str__(self) -> str:
...
+NamedRange: TypeAlias = tuple[Dimension, UnitRange]
+RelativeIndexElement: TypeAlias = IntIndex | slice | types.EllipsisType
+AbsoluteIndexElement: TypeAlias = NamedIndex | NamedRange
+AnyIndexElement: TypeAlias = RelativeIndexElement | AbsoluteIndexElement
+AbsoluteIndexSequence: TypeAlias = Sequence[NamedRange | NamedIndex]
+RelativeIndexSequence: TypeAlias = tuple[
+    slice | IntIndex | types.EllipsisType, ...
+]  # is a tuple but called Sequence for symmetry
+AnyIndexSequence: TypeAlias = RelativeIndexSequence | AbsoluteIndexSequence
+AnyIndex: TypeAlias = AnyIndexElement | AnyIndexSequence
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:171]: I took `AnyIndexSpec`

File: src/gt4py/next/common.py (hunk 39 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py]: The `[0]` means that only the first `dict` of the chainmap is taken, which is the `dict` added in that specific superclass. Taken the full chainmap of every superclass would create a wrong chainmap, because we will be concatenating the MROs of the superclasses instead of building up a chainmap with the dicts from the classes in the MRO of the new class.

File: src/gt4py/next/common.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py:337]: Ah, ok, good point. I would add a comment in the `DomainLike` definition mentioning it, just in case.

File: src/gt4py/next/common.py (hunk 39 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: Oh, makes sense.

File: src/gt4py/next/common.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:337]: I'll add it to the next PR, in case we don't touch this for other reasons (to save a few minutes waiting for ci)

File: src/gt4py/next/common.py (hunk 39 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: Should I take it in this PR?

File: src/gt4py/next/embedded/common.py (hunk 10 above)
Comment (egparedes) [reply src/gt4py/next/embedded/common.py:115]: I would add a message anyway, even it's basically just a comment.

File: src/gt4py/next/common.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py:337]: :+1:

File: src/gt4py/next/embedded/common.py (hunk 10 above)
Comment (havogt) [reply src/gt4py/next/embedded/common.py:115]: done

File: src/gt4py/next/common.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:337]: added now

File: src/gt4py/next/common.py (hunk 39 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py]: Let's do it in a different one.

**Review by egparedes — APPROVED** [review]
It looks good to me.
